pkg/policy/groups/helpers.go

```diff
@@ -0,0 +1,152 @@
+	clientset "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/policy/api"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	parentCNP   = "parentCNP"
+	cnpKindKey  = "CNPKind"
+	cnpKindName = "derivative"
+)
+
+var (
+	globalK8sClient *clientset.Clientset
+	k8sMutex        = lock.Mutex{}
+)
+
+// createDerivativeCNP will return a new CNP based on the given rule.
+func createDerivativeCNP(cnp *cilium_v2.CiliumNetworkPolicy) (*cilium_v2.CiliumNetworkPolicy, error) {
+	derivativeCNP := cnp.DeepCopy()
+	derivativeCNP.ObjectMeta.Name = fmt.Sprintf(
+		"%s-togroups-%s",
+		derivativeCNP.ObjectMeta.Name,
+		cnp.ObjectMeta.UID)
+	derivativeCNP.ObjectMeta.UID = ""
+	derivativeCNP.ObjectMeta.ResourceVersion = ""
+	derivativeCNP.ObjectMeta.Labels = map[string]string{
+		parentCNP:  string(cnp.ObjectMeta.UID),
+		cnpKindKey: cnpKindName,
+	}
+	derivativeCNP.Spec = &api.Rule{}
+	derivativeCNP.Specs = api.Rules{}
+
+	rules, err := cnp.Parse()
+	if err != nil {
+		return nil, fmt.Errorf("Cannot parse policies: %s", err)
+	}
+
+	for _, rule := range rules {
+		if !rule.RequiresDerivative() {
+			continue
+		}
+		newRule, err := rule.CreateDerivative()
+		if err != nil {
+			return derivativeCNP, err
+		}
+		derivativeCNP.Specs = append(derivativeCNP.Specs, newRule)
+	}
+
+	return derivativeCNP, nil
+}
+
+// getK8sClient return the kubernetes apiserver connection
+func getK8sClient() (*clientset.Clientset, error) {
+	k8sMutex.Lock()
+	defer k8sMutex.Unlock()
+	if globalK8sClient != nil {
+		return globalK8sClient, nil
+	}
+
+	restConfig, err := k8s.CreateConfig()
+	if err != nil {
+		return nil, fmt.Errorf("Unable to create rest configuration: %s", err)
+	}
+	k8sClient, err := clientset.NewForConfig(restConfig)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to create Kubernetes configuration: %s", err)
+
+	}
+	globalK8sClient = k8sClient
+	return globalK8sClient, nil
+}
+
+func updateCNPStatus(cnp *cilium_v2.CiliumNetworkPolicy) (*cilium_v2.CiliumNetworkPolicy, error) {
+	k8sClient, err := getK8sClient()
+	if err != nil {
+		// @TODO change the error here
+		return nil, err
+	}
+	return k8sClient.CiliumV2().CiliumNetworkPolicies(cnp.ObjectMeta.Namespace).UpdateStatus(cnp)
+}
+
+func updateOrCreateCNP(cnp *cilium_v2.CiliumNetworkPolicy) (*cilium_v2.CiliumNetworkPolicy, error) {
+	k8sClient, err := getK8sClient()
+	if err != nil {
+		return nil, err
+	}
+	k8sCNP, err := k8sClient.CiliumV2().CiliumNetworkPolicies(cnp.ObjectMeta.Namespace).
+		Get(cnp.ObjectMeta.Name, v1.GetOptions{})
+	if err == nil {
+		cnp.ObjectMeta.ResourceVersion = k8sCNP.ObjectMeta.ResourceVersion
+		return k8sClient.CiliumV2().CiliumNetworkPolicies(cnp.ObjectMeta.Namespace).Update(cnp)
+	}
+	return k8sClient.CiliumV2().CiliumNetworkPolicies(cnp.ObjectMeta.Namespace).Create(cnp)
+}
+
+func updateDerivativeStatus(cnp *cilium_v2.CiliumNetworkPolicy, derivativeName string, err error) error {
```

Comment:
We already have a function that updates status for a cnp with a lot of controllers, can't we use that one instead of doing new mistakes?

Reply:
could you point to me what is that function? I can't find it.

Reply:
I think aanm means this one (maybe) https://github.com/cilium/cilium/blob/master/daemon/k8s_watcher.go#L1325